File: sentil-java/src/main/java/io/github/sedislab/sentil/Trace.java
package io.github.sedislab.sentil;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

/** One or more named signals sampled over a strictly increasing sequence of times. */
public final class Trace extends NativeResource {
    Trace(long handle) {
        super(handle, NativeLib::traceDestroy);
    }

    /** A trace over the given times, with no signals yet. */
    public static Trace create(double[] times) throws SentilException {
        return new Trace(NativeLib.traceCreate(times));
    }

    /** A trace holding one named signal over the given times. */
    public static Trace fromSignal(double[] times, String name, double[] values)
            throws SentilException {
        return new Trace(NativeLib.traceFromSignal(times, name, values));
    }

    /** A trace with integer times 0, 1, ..., length - 1 and no signals yet. */
    public static Trace indexed(long length) throws SentilException {
        return new Trace(NativeLib.traceIndexed(length));
    }

    /** A trace parsed from CSV text, with the time column auto-detected. */
    public static Trace fromCsv(String text) throws SentilException {
        return new Trace(NativeLib.traceFromCsv(text));
    }

    /** A trace parsed from tab-separated text. */
    public static Trace fromTsv(String text) throws SentilException {
        return new Trace(NativeLib.traceFromTsv(text));
    }

    /** A trace read from a file, dispatching on the extension. */
    public static Trace fromPath(String path) throws SentilException {
        return new Trace(NativeLib.traceFromPath(path));
    }

    /** Add or replace a named signal, whose length must equal the trace length. */
    public void addSignal(String name, double[] values) throws SentilException {
        NativeLib.traceAddSignal(handle(), name, values);
    }

    /** The number of time points. */
    public long length() {
        return NativeLib.traceLen(handle());
    }

    /** Whether the trace has no time points. */
    public boolean isEmpty() {
        return NativeLib.traceIsEmpty(handle());
    }

    /** A copy of the sample times. */
    public double[] times() {
        return NativeLib.traceTimes(handle());
    }

    /** The names of the signals the trace carries, sorted. */
    public List<String> variables() throws SentilException {
        return Arrays.asList(NativeLib.traceVariables(handle()));
    }

    /** A copy of a named signal's values, or empty if the trace has no such signal. */
    public Optional<double[]> signal(String name) {
        return Optional.ofNullable(NativeLib.traceSignal(handle(), name));
    }

    /** A new trace resampled onto the given times with the given interpolation. */
    public Trace resample(double[] times, Interpolation interpolation) throws SentilException {
        return new Trace(NativeLib.traceResample(handle(), times, interpolation.code()));
    }

    /** Fix this trace's interpolation coefficients once. */
    public PreparedTrace prepare(Interpolation interpolation) throws SentilException {
        return new PreparedTrace(NativeLib.tracePrepare(handle(), interpolation.code()));
    }
}

File: sentil-matlab/+sentil/Trace.m
classdef Trace < handle
    %TRACE A multivariate timed signal.

    properties (SetAccess = private, Hidden)
        Handle uint64 = uint64(0)
    end

    methods (Static)
        function t = from_csv(text)
            %FROM_CSV Parse a trace from comma-separated text with a header row.
            t = sentil.Trace(sentil_mex('trace_from_csv', text));
        end

        function t = from_tsv(text)
            %FROM_TSV Parse a trace from tab-separated text with a header row.
            t = sentil.Trace(sentil_mex('trace_from_tsv', text));
        end

        function t = from_path(path)
            %FROM_PATH Load a trace from a CSV or TSV file, chosen by extension.
            t = sentil.Trace(sentil_mex('trace_from_path', path));
        end
    end

    methods
        function obj = Trace(varargin)
            %TRACE Build a trace from a time vector, optionally with one named signal.
            if nargin == 1 && isa(varargin{1}, 'uint64') && isscalar(varargin{1})
                obj.Handle = varargin{1};
            elseif nargin == 1
                obj.Handle = sentil_mex('trace_create', double(varargin{1}));
            elseif nargin == 3
                obj.Handle = sentil_mex('trace_from_signal', double(varargin{1}), ...
                    char(varargin{2}), double(varargin{3}));
            else
                error('sentil:trace', ...
                    'use sentil.Trace(times) or sentil.Trace(times, name, values)');
            end
        end

        function delete(obj)
            if obj.Handle ~= 0
                sentil_mex('trace_destroy', obj.Handle);
                obj.Handle = uint64(0);
            end
        end

        function add_signal(obj, name, values)
            %ADD_SIGNAL Attach a named signal sampled on the trace's time grid.
            obj.assertOpen();
            sentil_mex('trace_add_signal', obj.Handle, char(name), double(values));
        end

        function n = length(obj)
            %LENGTH The number of samples.
            obj.assertOpen();
            n = sentil_mex('trace_len', obj.Handle);
        end

        function tf = is_empty(obj)
            %IS_EMPTY Whether the trace has no samples.
            obj.assertOpen();
            tf = sentil_mex('trace_is_empty', obj.Handle);
        end

        function t = times(obj)
            %TIMES The time grid as a row vector.
            obj.assertOpen();
            t = sentil_mex('trace_times', obj.Handle);
        end

        function v = variables(obj)
            %VARIABLES The signal names.
            obj.assertOpen();
            v = sentil_mex('trace_variables', obj.Handle);
        end

        function s = signal(obj, name)
            %SIGNAL The named signal, or [] if the trace has no such name.
            obj.assertOpen();
            s = sentil_mex('trace_signal', obj.Handle, char(name));
        end

        function t = resample(obj, newTimes, interp)
            %RESAMPLE A copy of the trace read onto a new time grid.
            obj.assertOpen();
            if nargin < 3
                interp = sentil.Interpolation.Linear;
            end
            t = sentil.Trace(sentil_mex('trace_resample', obj.Handle, double(newTimes), ...
                double(int32(interp))));
        end

        function p = prepare(obj, interp)
            %PREPARE A sentil.PreparedTrace with precomputed interpolation.
            obj.assertOpen();
            if nargin < 2
                interp = sentil.Interpolation.Linear;
            end
            p = sentil.PreparedTrace(uint64(sentil_mex('trace_prepare', obj.Handle, ...
                double(int32(interp)))));
        end
    end

    methods (Access = private)
        function assertOpen(obj)
            if obj.Handle == 0
                error('sentil:handle', 'this trace has been closed');
            end
        end
    end
end